### PR TITLE
Modify semantics for zero-sized allocations in  `axom::reallocate()`

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This PR modifies the semantics of ``axom::reallocate(_, 0)`` to always return a valid pointer
- This matches the semantics of ``Umpire``'s ``reallocate(_,0)`` function (after a bugfix in https://github.com/LLNL/Umpire/pull/292)
   - This PR includes a workaround to achieve this new behavior with our existing version of Umpire. We can consider removing this workaround after we upgrade Umpire to a version later than v1.1.
- Also updates sidre's test to verify a ``Buffer`` or ``View`` is allocated after a zero-sized allocation is requested
  - Note that there were no changes necessary to ``sidre``, only to our sidre tests related to zero-sized allocations.
- This change fixes a regression for Visualizing mesh blueprint files in a domain decomposed case where some domains did not have elements for a particular mesh. 
   - For more details, see: https://github.com/visit-dav/visit/issues/4212

# Design review)

- We discussed this on the ``axom-dev`` mailing list and in a team meeting on 1/6/2020.
- This issue was also brought up with several customers during a meeting on 1/8/2020.

### CODE REVIEW

File: src/axom/core/memory_management.hpp [hunk 1]
```diff
@@ -200,23 +194,47 @@ inline void deallocate( T*& pointer ) noexcept
...
-  {
-    axom::deallocate( pointer );
-    return nullptr;
-  }
-
   const std::size_t numbytes = n * sizeof( T );
 
 #ifdef AXOM_USE_UMPIRE
 
+  // Workaround for bug in Umpire's handling on reallocate(0)
```
Comment (kennyweiss) [inline src/axom/core/memory_management.hpp]: Note: These workarounds will be necessary until we and all our customers build against an as-yet-unreleased version of Umpire (post v1.1).

Should we guard these with a config option? E.g. ``AXOM_USE_UMPIRE_REALLOC_WORKAROUND`` which will be ``ON`` by default until we upgrade Umpire and then ``OFF`` by default?

File: src/axom/core/memory_management.hpp [hunk 2]
```diff
@@ -200,23 +194,47 @@ inline void deallocate( T*& pointer ) noexcept
...
+  }
+
   pointer = static_cast< T* >( rm.reallocate( pointer, numbytes ) );
 
 #else
 
   pointer = static_cast< T* >( std::realloc( pointer, numbytes ) );
 
+  // Consistently handle realloc(0) for std::realloc to match umpire's behavior
+  if(n==0 && pointer == nullptr)
```
Comment (kennyweiss) [inline src/axom/core/memory_management.hpp]: Note: This ensures that the result of ``axom::reallocate(0)`` is the same, regardless of whether Axom is configured with Umpire.

**Comment by kennyweiss** [comment]
Tag: @rrsettgast, @corbett5, @mcfadden8

File: src/axom/core/memory_management.hpp (hunk 1 above)
Comment (rhornung67) [reply src/axom/core/memory_management.hpp]: Since it appears that an Umpire release is imminent (they started the release process yesterday, I believe), I say leave this as is and clean it up when we pull in the new Umpire.

File: src/axom/core/memory_management.hpp [hunk 3]
```diff
@@ -200,23 +194,47 @@ inline void deallocate( T*& pointer ) noexcept
...
-    return nullptr;
-  }
-
   const std::size_t numbytes = n * sizeof( T );
 
 #ifdef AXOM_USE_UMPIRE
 
+  // Workaround for bug in Umpire's handling on reallocate(0)
+  // Fixed in Umpire PR #292 (after v1.1.0)
+  if(n==0)
```
Comment (gzagaris) [inline src/axom/core/memory_management.hpp]: Please update the documentation of this method as well and document the behavior when `n==0` in a `\note`.

Comment (kennyweiss) [reply src/axom/core/memory_management.hpp]: Will do.

Comment (gzagaris) [reply src/axom/core/memory_management.hpp]: Thanks!

Comment (kennyweiss) [reply src/axom/core/memory_management.hpp]: Done.
